pydantic/types.py

```diff
@@ -1460,29 +1476,23 @@ def __eq__(self, other: Any) -> bool:
     def __hash__(self) -> int:
         return hash(self.get_secret_value())
 
-    def __len__(self) -> int:
-        return len(self._secret_value)
-
     def __str__(self) -> str:
         return str(self._display())
 
     def __repr__(self) -> str:
         return f'{self.__class__.__name__}({self._display()!r})'
 
-    def _display(self) -> SecretType:
+    def __len__(self) -> int:
+        if isinstance(self._secret_value, (str, bytes)):
+            return len(self._secret_value)
+        else:
+            raise TypeError(f'len() of {self.__class__.__name__} is not supported')
```

Comment:
`object of type 'datetime.date' has no len()` is the wording used when you do `len(date.today())`, we should copy that.